commit 7c4ce5a154954c78f026b6ff2367214cd802b431
Author: Talustus <talustus@dream-irc.de>
Date:   Sat May 25 18:51:09 2013 +0200

    GPU-OC: Update GPU-OC
      * make GPU Overclocking configurable via config switch

diff --git a/arch/arm/mach-msm/board-8064-gpu.c b/arch/arm/mach-msm/board-8064-gpu.c
index 080155f..d5689fe 100644
--- a/arch/arm/mach-msm/board-8064-gpu.c
+++ b/arch/arm/mach-msm/board-8064-gpu.c
@@ -138,13 +138,21 @@ static struct msm_bus_vectors grp3d_max_vectors[] = {
 		.src = MSM_BUS_MASTER_GRAPHICS_3D,
 		.dst = MSM_BUS_SLAVE_EBI_CH0,
 		.ab = 0,
+#ifdef CONFIG_KGSL_GPUOC
 		.ib = KGSL_CONVERT_TO_MBPS(4600),
+#else
+		.ib = KGSL_CONVERT_TO_MBPS(4264),
+#endif
 	},
 	{
 		.src = MSM_BUS_MASTER_GRAPHICS_3D_PORT1,
 		.dst = MSM_BUS_SLAVE_EBI_CH0,
 		.ab = 0,
+#ifdef CONFIG_KGSL_GPUOC
 		.ib = KGSL_CONVERT_TO_MBPS(4600),
+#else
+		.ib = KGSL_CONVERT_TO_MBPS(4264),
+#endif
 	},
 };
 
@@ -220,6 +228,7 @@ static struct kgsl_device_iommu_data kgsl_3d0_iommu_data[] = {
 
 static struct kgsl_device_platform_data kgsl_3d0_pdata = {
 	.pwrlevel = {
+#ifdef CONFIG_KGSL_GPUOC
 		{
 			.gpu_freq = 533000000,
 			.bus_freq = 4,
@@ -230,6 +239,7 @@ static struct kgsl_device_platform_data kgsl_3d0_pdata = {
 			.bus_freq = 4,
 			.io_fraction = 0,
 		},
+#endif
 		{
 			.gpu_freq = 400000000,
 			.bus_freq = 4,
@@ -256,7 +266,11 @@ static struct kgsl_device_platform_data kgsl_3d0_pdata = {
 		},
 	},
 	.init_level = 1,
+#ifdef CONFIG_KGSL_GPUOC
 	.num_levels = 7,
+#else
+	.num_levels = 5,
+#endif
 	.set_grp_async = NULL,
 	.idle_timeout = HZ/10,
 	.nap_allowed = true,
@@ -287,7 +301,11 @@ void __init apq8064_init_gpu(void)
 	unsigned int version = socinfo_get_version();
 
 	if (cpu_is_apq8064ab())
+#ifdef CONFIG_KGSL_GPUOC
 		kgsl_3d0_pdata.pwrlevel[0].gpu_freq = 533000000;
+#else
+		kgsl_3d0_pdata.pwrlevel[0].gpu_freq = 450000000;
+#endif
 	if (SOCINFO_VERSION_MAJOR(version) == 2) {
 		kgsl_3d0_pdata.chipid = ADRENO_CHIPID(3, 2, 0, 2);
 	} else {
diff --git a/arch/arm/mach-msm/clock-8960.c b/arch/arm/mach-msm/clock-8960.c
index 7b93904..e27752a 100644
--- a/arch/arm/mach-msm/clock-8960.c
+++ b/arch/arm/mach-msm/clock-8960.c
@@ -597,7 +597,7 @@ static struct pll_clk pll15_clk = {
 	.parent = &pxo_clk.c,
 	.c = {
 		.dbg_name = "pll15_clk",
-		.rate = 900000000,
+		.rate = 975000000,
 		.ops = &clk_ops_local_pll,
 		CLK_INIT(pll15_clk.c),
 	},
@@ -3531,7 +3531,9 @@ static struct clk_freq_tbl clk_tbl_gfx3d[] = {
 	F_GFX3D(320000000, pll2,  2,  5),
 	F_GFX3D(400000000, pll2,  1,  2),
 	F_GFX3D(450000000, pll15, 1,  2),
+#ifdef CONFIG_KGSL_GPUOC
 	F_GFX3D(533000000, pll15, 1,  2),
+#endif
 	F_END
 };
 
@@ -3553,8 +3555,10 @@ static struct clk_freq_tbl clk_tbl_gfx3d_8960[] = {
 	F_GFX3D(300000000, pll3, 1,  4),
 	F_GFX3D(320000000, pll2, 2,  5),
 	F_GFX3D(400000000, pll2, 1,  2),
+#ifdef CONFIG_KGSL_GPUOC
 	F_GFX3D(450000000, pll15, 1,  2),
 	F_GFX3D(533000000, pll15, 1,  2),
+#endif
 	F_END
 };
 
@@ -3576,39 +3580,63 @@ static struct clk_freq_tbl clk_tbl_gfx3d_8930ab[] = {
 	F_GFX3D(266667000, pll2,  1,  3),
 	F_GFX3D(320000000, pll2,  2,  5),
 	F_GFX3D(400000000, pll2,  1,  2),
+#ifdef CONFIG_KGSL_GPUOC
 	F_GFX3D(450000000, pll15, 1,  2),
 	F_GFX3D(533000000, pll15, 1,  2),
+#else
+	F_GFX3D(500000000, pll15, 1,  2),
+#endif
 	F_END
 };
 
 static unsigned long fmax_gfx3d_8064ab[VDD_DIG_NUM] = {
 	[VDD_DIG_LOW]     = 128000000,
 	[VDD_DIG_NOMINAL] = 325000000,
+#ifdef CONFIG_KGSL_GPUOC
 	[VDD_DIG_HIGH]    = 533000000
+#else
+	[VDD_DIG_HIGH]    = 450000000
+#endif
 };
 
 static unsigned long fmax_gfx3d_8064[VDD_DIG_NUM] = {
 	[VDD_DIG_LOW]     = 128000000,
 	[VDD_DIG_NOMINAL] = 325000000,
+#ifdef CONFIG_KGSL_GPUOC
 	[VDD_DIG_HIGH]    = 533000000
+#else
+	[VDD_DIG_HIGH]    = 400000000
+#endif
 };
 
 static unsigned long fmax_gfx3d_8930[VDD_DIG_NUM] = {
 	[VDD_DIG_LOW]     = 192000000,
 	[VDD_DIG_NOMINAL] = 320000000,
+#ifdef CONFIG_KGSL_GPUOC
 	[VDD_DIG_HIGH]    = 533000000
+#else
+	[VDD_DIG_HIGH]    = 400000000
+#endif
 };
 
 static unsigned long fmax_gfx3d_8930aa[VDD_DIG_NUM] = {
 	[VDD_DIG_LOW]     = 192000000,
 	[VDD_DIG_NOMINAL] = 320000000,
+#ifdef CONFIG_KGSL_GPUOC
 	[VDD_DIG_HIGH]    = 533000000
+#else
+	[VDD_DIG_HIGH]    = 450000000
+#endif
 };
 
 static unsigned long fmax_gfx3d_8930ab[VDD_DIG_NUM] = {
 	[VDD_DIG_LOW]     = 192000000,
 	[VDD_DIG_NOMINAL] = 320000000,
+#ifdef CONFIG_KGSL_GPUOC
 	[VDD_DIG_HIGH]    = 533000000
+#else
+	[VDD_DIG_HIGH]    = 500000000
+#endif
 };
 
 static struct bank_masks bmnd_info_gfx3d = {
@@ -3650,7 +3678,11 @@ static struct rcg_clk gfx3d_clk = {
 		.dbg_name = "gfx3d_clk",
 		.ops = &clk_ops_rcg,
 		VDD_DIG_FMAX_MAP3(LOW,  128000000, NOMINAL, 300000000,
+#ifdef CONFIG_KGSL_GPUOC
 				  HIGH, 533000000),
+#else
+				  HIGH, 400000000),
+#endif
 		CLK_INIT(gfx3d_clk.c),
 		.depends = &gmem_axi_clk.c,
 	},
@@ -6682,7 +6714,7 @@ static void __init msm8960_clock_pre_init(void)
 	else if (cpu_is_msm8930aa())
 		gfx3d_clk.c.fmax = fmax_gfx3d_8930aa;
 	if (cpu_is_msm8930() || cpu_is_msm8930aa() || cpu_is_msm8627()) {
-		pll15_clk.c.rate = 975000000;
+		pll15_clk.c.rate = 900000000;
 		gmem_axi_clk.c.depends = &gfx3d_axi_clk_8930.c;
 	} else if (cpu_is_msm8930ab()) {
 		gfx3d_clk.freq_tbl = clk_tbl_gfx3d_8930ab;
diff --git a/arch/arm/mach-msm/include/mach/kgsl.h b/arch/arm/mach-msm/include/mach/kgsl.h
index 01f1087..47007da 100644
--- a/arch/arm/mach-msm/include/mach/kgsl.h
+++ b/arch/arm/mach-msm/include/mach/kgsl.h
@@ -21,7 +21,11 @@
 #define KGSL_CLK_MEM_IFACE 0x00000010
 #define KGSL_CLK_AXI	0x00000020
 
+#ifdef CONFIG_KGSL_GPUOC
 #define KGSL_MAX_PWRLEVELS 7
+#else
+#define KGSL_MAX_PWRLEVELS 5
+#endif
 
 #define KGSL_CONVERT_TO_MBPS(val) \
 	(val*1000*1000U)
diff --git a/drivers/gpu/msm/Kconfig b/drivers/gpu/msm/Kconfig
index ba63fbc..e965e94 100644
--- a/drivers/gpu/msm/Kconfig
+++ b/drivers/gpu/msm/Kconfig
@@ -96,3 +96,8 @@ config MSM_KGSL_DISABLE_SHADOW_WRITES
 	bool "Disable register shadow writes for context switches"
 	default n
 	depends on MSM_KGSL
+
+config KGSL_GPUOC
+	bool "Enable MSM KGSL GPU Overclocking to 533MHz"
+	default n
+	depends on MSM_KGSL
